eth/client.go

```diff
@@ -150,306 +226,601 @@ func NewTransactOptsForAccount(account accounts.Account, passphrase string, keyS
 
-	return c.backend.SubscribeFilterLogs(ctx, q, logsCh)
+		tx, err := c.roundsManagerSession.InitializeRound()
+		if err != nil {
+			outErr <- err
+			return
+		}
+
+		glog.Infof("[%v] Submitted tx %v. Initialize round", c.account.Address.Hex(), tx.Hash().Hex())
+
+		receipt, err := c.WaitForReceipt(tx)
+		if err != nil {
+			outErr <- err
+		} else {
+			outRes <- *receipt
+		}
+
+		return
+	}()
+
+	return outRes, outErr
 }
 
-func monitorJobEvent(logsCh <-chan types.Log, callback func(l types.Log) error) {
-	for {
-		select {
-		case log := <-logsCh:
-			if !log.Removed && log.BlockNumber != 0 {
-				// TODO: Perform action in response to job event
-				// glog.Infof("Received a job at block %v with tx %v by address %v", log.BlockNumber, log.TxHash.Hex(), log.Address.Hex())
-				callback(log)
-			}
+func (c *Client) Transcoder(blockRewardCut uint8, feeShare uint8, pricePerSegment *big.Int) (<-chan types.Receipt, <-chan error) {
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
+
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
+
+		tx, err := c.bondingManagerSession.Transcoder(blockRewardCut, feeShare, pricePerSegment)
+		if err != nil {
+			outErr <- err
+			return
 		}
-	}
+
+		glog.Infof("[%v] Submitted tx %v. Register as transcoder", c.account.Address.Hex(), tx.Hash().Hex())
+
+		receipt, err := c.WaitForReceipt(tx)
+		if err != nil {
+			outErr <- err
+			return
+		}
+
+		outRes <- *receipt
+
+		return
+	}()
+
+	return outRes, outErr
 }
 
-func (c *Client) WatchEvent(logsCh <-chan types.Log) (types.Log, error) {
-	var (
-		timer = time.NewTimer(c.eventTimeout)
-	)
+func (c *Client) Bond(amount *big.Int, toAddr common.Address) (<-chan types.Receipt, <-chan error) {
+	inRes, inErr := c.Approve(c.bondingManagerAddr, amount)
+
+	timer := time.NewTimer(c.eventTimeout)
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
+
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
 
-	for {
 		select {
-		case log := <-logsCh:
+		case log := <-inRes:
 			if !log.Removed {
-				return log, nil
+				tx, err := c.bondingManagerSession.Bond(amount, toAddr)
+				if err != nil {
+					outErr <- err
+					return
+				}
+
+				glog.Infof("[%v] Submitted tx %v. Bond %v LPTU to %v", c.account.Address.Hex(), tx.Hash().Hex(), amount, toAddr.Hex())
+
+				receipt, err := c.WaitForReceipt(tx)
+				if err != nil {
+					outErr <- err
+				} else {
+					outRes <- *receipt
+				}
+
+				return
 			}
+		case err := <-inErr:
+			outErr <- err
+			return
 		case <-timer.C:
-			err := fmt.Errorf("watchEvent timed out")
+			outErr <- fmt.Errorf("Event subscription timed out")
+			return
+		}
+	}()
+
+	return outRes, outErr
+}
+
+func (c *Client) Reward() (<-chan types.Receipt, <-chan error) {
+	outRes := make(chan types.Receipt)
+	outErr := make(chan error)
 
-			glog.Errorf(err.Error())
-			return types.Log{}, err
+	go func() {
+		defer close(outRes)
+		defer close(outErr)
+
+		tx, err := c.bondingManagerSession.Reward()
```

Comment:
There is quite a bit of repeated code between the different protocol transaction invocations: waiting for receipts, checking for errors, and setting the error and return values. I know each function does have different args and slightly different logging and semantics so it may be hard to abstract some of it out, but maybe @ericxtang has an idea for how to do this in go using function pointers or some other pattern?